File: fullstacks/rails/app/services/favorite_service.rb
module FavoriteService
  module_function

  def add_favorite(user, capsule_id_raw)
    capsule_id = CrossDb.parse_uuid_or_nil(capsule_id_raw) or raise ApiError.not_found("胶囊不存在")

    # favorite_count 是冗余计数器，必须与 favorites 行变更同处一个事务。
    # Postgres 路径取胶囊行写锁；SQLite 依赖单写事务与原子提交。
    ActiveRecord::Base.transaction do
      rel = Capsule.where(id: capsule_id)
      rel = rel.lock unless AppConfig.sqlite?
      capsule = rel.first
      raise ApiError.not_found("胶囊不存在") if capsule.nil? || !capsule.in_plaza
      raise ApiError.bad_request("不能收藏自己创建的胶囊") if capsule.owner_id == user.id

      existing = Favorite.find_by(user_id: user.id, capsule_id: capsule.id)
      if existing
        next { capsuleId: capsule.id.to_s, favoriteCount: capsule.favorite_count,
               favoritedAt: CrossDb.iso_instant(existing.created_at) }
      end

      now = Time.now.utc
      Favorite.create!(user_id: user.id, capsule_id: capsule.id, created_at: now)
      Capsule.where(id: capsule.id).update_all("favorite_count = favorite_count + 1")
      new_count = Capsule.where(id: capsule.id).pick(:favorite_count) || (capsule.favorite_count + 1)
      { capsuleId: capsule.id.to_s, favoriteCount: new_count, favoritedAt: CrossDb.iso_instant(now) }
    end
  end

  def remove_favorite(user, capsule_id_raw)
    # 取消收藏幂等：胶囊不存在 / 格式非法 / 原本未收藏都返回成功（204）。
    capsule_id = CrossDb.parse_uuid_or_nil(capsule_id_raw)
    return if capsule_id.nil?

    ActiveRecord::Base.transaction do
      rel = Capsule.where(id: capsule_id)
      rel = rel.lock unless AppConfig.sqlite?
      capsule = rel.first
      next if capsule.nil?

      if Favorite.exists?(user_id: user.id, capsule_id: capsule.id)
        Favorite.where(user_id: user.id, capsule_id: capsule.id).delete_all
        Capsule.where(id: capsule.id).where("favorite_count > 0")
               .update_all("favorite_count = favorite_count - 1")
      end
    end
  end
end

File: fullstacks/laravel/app/Services/FavoriteService.php
<?php

namespace App\Services;

use App\Exceptions\ApiError;
use App\Models\Capsule;
use App\Models\Favorite;
use App\Models\User;
use App\Support\CrossDb;
use App\Support\Mapper;
use Illuminate\Support\Facades\DB;

/**
 * 收藏：添加（事务 + 行锁防 favorite_count 漂移、禁止收藏自己/未公开胶囊）、
 * 取消（幂等：非法/不存在/未收藏均按成功处理）、"我收藏的"列表。数据访问走 Eloquent。
 */
class FavoriteService
{
    public function __construct(
        private readonly CrossDb $db,
        private readonly Mapper $mapper,
    ) {
    }

    public function myFavorites(User $user, array $query): array
    {
        [$page, $size] = $this->mapper->pageParams($query);
        $total = Favorite::where('user_id', $user->id)->count();
        $favorites = Favorite::with('capsule.owner')
            ->where('user_id', $user->id)
            ->orderByDesc('created_at')
            ->forPage($page, $size)
            ->get();

        $items = $favorites
            // 复刻原 INNER JOIN capsules/users 语义：胶囊或其 owner 缺失的收藏行被排除。
            ->filter(fn (Favorite $f) => $f->capsule !== null && $f->capsule->owner !== null)
            ->map(fn (Favorite $f) => $this->mapper->listItem($f->capsule, $user->id, $f->created_at))
            ->values()
            ->all();

        return ['items' => $items, 'pagination' => $this->mapper->pagination($page, $size, $total)];
    }

    public function addFavorite(User $user, string $capsuleId): array
    {
        $canonical = $this->db->canonicalUuid($capsuleId);
        if ($canonical === null) throw new ApiError(404, 'NOT_FOUND', '胶囊不存在');
        $dbId = $this->db->idToDb($canonical);

        return DB::transaction(function () use ($user, $canonical, $dbId) {
            // 行锁住胶囊，序列化对 favorite_count 的并发更新。
            $cap = Capsule::where('id', $dbId)
                ->where('in_plaza', $this->db->boolParam(true))
                ->lockForUpdate()
                ->first();
            if (!$cap) throw new ApiError(404, 'NOT_FOUND', '胶囊不存在');
            if ($cap->owner_id === $user->id) throw new ApiError(400, 'BAD_REQUEST', '不能收藏自己的胶囊');

            $existing = Favorite::where('user_id', $user->id)->where('capsule_id', $dbId)->first();
            if ($existing) {
                return ['capsuleId' => $canonical, 'favoriteCount' => (int) $cap->favorite_count, 'favoritedAt' => $this->db->iso($existing->created_at)];
            }

            $now = $this->db->nowDb();
            Favorite::create(['user_id' => $user->id, 'capsule_id' => $dbId, 'created_at' => $now]);
            $cap->increment('favorite_count', 1, ['updated_at' => $now]);

            return ['capsuleId' => $canonical, 'favoriteCount' => (int) $cap->favorite_count, 'favoritedAt' => $this->db->iso($now)];
        });
    }

    public function removeFavorite(User $user, string $capsuleId): void
    {
        // 取消收藏幂等：格式非法 / 胶囊不存在 / 原本未收藏都返回成功（204）。
        $canonical = $this->db->canonicalUuid($capsuleId);
        if ($canonical === null) return;
        $dbId = $this->db->idToDb($canonical);

        DB::transaction(function () use ($user, $dbId) {
            Capsule::where('id', $dbId)->lockForUpdate()->first();
            $deleted = Favorite::where('user_id', $user->id)->where('capsule_id', $dbId)->delete();
            if ($deleted > 0) {
                // WHERE favorite_count > 0 即把计数下限钳在 0（只在 >0 时才减 1）。
                Capsule::where('id', $dbId)
                    ->where('favorite_count', '>', 0)
                    ->decrement('favorite_count', 1, ['updated_at' => $this->db->nowDb()]);
            }
        });
    }
}
